### PR TITLE
add user supplied function when parse()

### PR DESCRIPTION
**Pull request by xilabao**
Allow user to get the flag/value in the order. The unit test example try to restore the os.Args. (cannot restore the args abandoned)

### CODE REVIEW

File: flag_test.go [hunk 1]
```diff
@@ -333,6 +333,95 @@ func testParse(f *FlagSet, t *testing.T) {
 	}
 }
 
+func testParseAll(f *FlagSet, t *testing.T) {
```
Comment (moorereason) [inline flag_test.go]: This test func is never called.

File: flag.go [hunk 2]
```diff
@@ -881,7 +887,29 @@ func (f *FlagSet) parseArgs(args []string) (err error) {
 func (f *FlagSet) Parse(arguments []string) error {
 	f.parsed = true
 	f.args = make([]string, 0, len(arguments))
-	err := f.parseArgs(arguments)
+	err := f.parseArgs(arguments, nil)
```
Comment (eparis) [inline flag.go]: What do you think about defining a function here which does the:
```
err = f.setFlag(flag, value, shorthands)
```
So that `ParseAll()` when called with a custom function will have no implied side effects.

It also means we wouldn't ever have to test for fn == nil.

File: flag_test.go (hunk 1 above)
Comment (xilabao) [inline flag_test.go]: thanks, fixed

File: flag.go (hunk 2 above)
Comment (xilabao) [inline flag.go]: parseShortArg() need to handle "-flag1flag2",  I guess that it need to parse the flag again then to do `f.setFlag`.

Comment (eparis) [inline flag.go]: Make the user function take `(flag, value, origArg)` ? Or even don't pass `origArg` at all to the user functionl? When you 'rebuild' your command line using this function it would translate `-abcd` into `-a -b -c -d` and `-oarg` into `-o arg` but who cares?

Comment (xilabao) [inline flag.go]: @eparis I change it on your suggestion. If people didn't give the value when the flag used, will return an empty string for value.  They can use Lookup/LookupByShorthand to get the value. Is it OK?

**Comment by xilabao** [comment]
@eparis @moorereason PTAL

**Comment by eparis** [comment]
What do you think of #108 ? Its basically your exact same thing, but running Parse a second time with a custom function is non-destructive...

**Comment by xilabao** [comment]
@eparis Sorry to bother you. The feature has been forgotten for a while. I update it at your suggestion. Please take a took again.

**Comment by eparis** [comment]
I really like the code.  CI failed. I'm worried `go test` is also going to fail because of some error handling, but maybe I'm wrong about that.

**Comment by xilabao** [comment]
Fixed. change `ParseFunc` to `parseFunc`. @eparis
